src/mds/Server.cc

```diff
@@ -2079,6 +2079,39 @@ void Server::handle_slave_auth_pin_ack(MDRequestRef& mdr, MMDSSlaveRequest *ack)
 // HELPERS
 
 
+/**
+ * check whether we are permitted to complete a request
+ *
+ * Check whether we have permission to perform the operation specified
+ * by mask on the given inode, based on the capability in the mdr's
+ * session.
+ */
+bool Server::check_access(MDRequestRef& mdr, CInode *in, unsigned mask)
+{
+  Session *s = mdr->session;
+
+  // FIXME: populate these with real values
+  string path;
+  int uid = mdr->client_request->get_caller_uid();
+
+  // simple read/write check
+  for (vector<MDSCapGrant>::iterator p = s->auth_caps.grants.begin();
+       p != s->auth_caps.grants.end();
+       ++p) {
+    if (p->match.match(path, uid)) {
+      if (p->spec.allows(mask & MAY_READ,
+			 mask & (MAY_WRITE|MAY_EXECUTE))) {
+	return true;
```

Comment:
Because each grant specifies/may specify a uid and gids:

 allow rw uid 123 gids 123,1000
 allow r uid 456 gids 456,2000

Depending on which one matches, we need to take that uid and gid list and check unix permissions.  Either we need to make callbacks out of the MDSAuthCap struct, or we need to expose the fact that it is a list of distinct grants.

The fact that it's a list is no secret, so my preference is the latter.  We can make an accessor to get at the vector<> iterator (that is, say, const) if that makes you feel better?